File: soft2/lec07/python/echo-server.py
#!/usr/bin/env python

# http://omniorb.sourceforge.net/omnipy3/omniORBpy/omniORBpy002.html
import sys
from omniORB import CORBA, PortableServer
import CosNaming, EchoApp, EchoApp__POA

# Define an implementation of the Echo interface
class Echo_i (EchoApp__POA.Echo):
    def echoString(self, mesg):
        print "echoString() called with message:", mesg
        return mesg + ", Copy"

# Initialise the ORB and find the root POA
orb = CORBA.ORB_init(sys.argv, CORBA.ORB_ID)
poa = orb.resolve_initial_references("RootPOA")

# Create an instance of Echo_i and an Echo object reference
ei = Echo_i()
eo = ei._this()

# Obtain a reference to the root naming context
obj         = orb.resolve_initial_references("NameService")
rootContext = obj._narrow(CosNaming.NamingContext)

if rootContext is None:
    print "Failed to narrow the root naming context"
    sys.exit(1)

# Bind a context named "test.my_context" to the root context
name = [CosNaming.NameComponent("contest", "")]
try:
    testContext = rootContext.bind_new_context(name)
    print "New test context bound"

except CosNaming.NamingContext.AlreadyBound, ex:
    print "Test context already exists"
    obj = rootContext.resolve(name)
    testContext = obj._narrow(CosNaming.NamingContext)
    if testContext is None:
        print "test.mycontext exists but is not a NamingContext"
        sys.exit(1)

# Bind the Echo object to the test context
name = [CosNaming.NameComponent("Example", "")]
try:
    testContext.bind(name, eo)
    print "New Example object bound"

except CosNaming.NamingContext.AlreadyBound:
    testContext.rebind(name, eo)
    print "Example binding already existed -- rebound"

# Activate the POA
poaManager = poa._get_the_POAManager()
poaManager.activate()

# Block for ever (or until the ORB is shut down)
orb.run()

File: soft2/lec07/cpp/echo-server.cpp
// https://xennis.org/wiki/CORBA#Server

#include "echo.hh"
#include "echo_impl.h"
#include <iostream>

/** Server name, clients needs to know this name */
#define SERVER_NAME		"Example"
using namespace std;

int main(int argc, char ** argv)
{
  try {
    CORBA::ORB_ptr orb = CORBA::ORB_init(argc, argv);
    CORBA::Object_var poa_obj = orb->resolve_initial_references("RootPOA");
    PortableServer::POA_var poa = PortableServer::POA::_narrow(poa_obj);
    PortableServer::POAManager_var manager = poa->the_POAManager();
    //------------------------------------------------------------------------
    // Create service
    //------------------------------------------------------------------------
    EchoApp_impl * service = new EchoApp_impl;
    try {
      CORBA::Object_var ns_obj = orb->resolve_initial_references("NameService");
      if (!CORBA::is_nil(ns_obj)) {
	CosNaming::NamingContext_ptr nc = CosNaming::NamingContext::_narrow(ns_obj);
	CosNaming::Name context_name;
	context_name.length(1);
	context_name[0].id = CORBA::string_dup("context");
	context_name[0].kind = CORBA::string_dup("");
	CosNaming::NamingContext_var tc;
	try{
	  tc = nc->bind_new_context(context_name);
	}catch(CosNaming::NamingContext::AlreadyBound& ex) {
	  CORBA::Object_var obj;
	  obj = nc->resolve(context_name);
	  tc = CosNaming::NamingContext::_narrow(obj);
	  if ( CORBA::is_nil(tc) ) {
	    cerr << "Failed to narrow naming context." << endl;
	    return 0;
	  }
	}
	CosNaming::Name object_name;
	object_name.length(1);
	object_name[0].id = CORBA::string_dup(SERVER_NAME);
	object_name[0].kind = CORBA::string_dup("");
	try{
	  tc->bind(object_name, service->_this());
	}catch(CosNaming::NamingContext::AlreadyBound& ex) {
	  cerr << "Eample binding already existed -- rebound" << endl;
	  tc->rebind(object_name, service->_this());
	}
	cout << argv[0] << " C++ (omniORB) server '" << SERVER_NAME << "' is running .." << endl;
      }
    } catch (CosNaming::NamingContext::NotFound &) {
      cerr << "Caught CORBA exception: not found" << endl;
    } catch (CosNaming::NamingContext::InvalidName &) {
      cerr << "Caught CORBA exception: invalid name" << endl;
    } catch (CosNaming::NamingContext::CannotProceed &) {
      cerr << "Caught CORBA exception: cannot proceed" << endl;
    }
    manager->activate();
    orb->run();
    delete service;
    orb->destroy();
  } catch (CORBA::UNKNOWN) {
    cerr << "Caught CORBA exception: unknown exception" << endl;
  } catch (CORBA::SystemException &) {
    cerr << "Caught CORBA exception: system exception" << endl;
  }
}
